File: scripts/linters/contiguous_use_block.rb
def contiguous_use_block(root_dir, excludes = [])
  pattern = root_dir.join('crates', '**', '*.rs').to_s
  errors = Dir.glob(pattern).sort.flat_map do |path|
    relative = Pathname.new(path).relative_path_from(root_dir).to_s
    next [] if excludes.include?(relative)

    find_split_use_block(path, relative)
  end

  return true if errors.empty?

  puts 'Found blank lines splitting the leading `use` block into sections.'
  puts 'All `use` statements at the top of the file must be contiguous (no blank lines between them):'
  errors.each do |err|
    puts "  #{err}"
  end
  false
end

USE_START_RE = /\A(?:pub(?:\([^)]*\))?\s+)?use\b/

def find_split_use_block(path, relative)
  lines = File.readlines(path)
  errors = []

  i = skip_preamble(lines)
  return [] if i.nil?

  loop do
    i = consume_use_statement(lines, i)
    break if i >= lines.length

    blanks = []
    j = i
    while j < lines.length
      stripped = lines[j].strip
      if stripped.empty?
        blanks << j
        j += 1
      elsif stripped.start_with?('//') || stripped.start_with?('#[')
        j += 1
      else
        break
      end
    end

    if j < lines.length && lines[j].strip =~ USE_START_RE
      blanks.each do |bi|
        errors << "#{relative}:#{bi + 1}: blank line splits the leading `use` block"
      end
      i = j
    else
      break
    end
  end

  errors
end

def skip_preamble(lines)
  lines.each_with_index do |raw, idx|
    stripped = raw.strip
    return idx if stripped =~ USE_START_RE
    next if stripped.empty? || stripped.start_with?('//') || stripped.start_with?('#![') || stripped.start_with?('#[')

    return nil
  end
  nil
end

def consume_use_statement(lines, start_idx)
  brace_depth = 0
  i = start_idx
  while i < lines.length
    line = lines[i]
    brace_depth += line.count('{') - line.count('}')
    done = brace_depth <= 0 && line.rstrip.end_with?(';')
    i += 1
    return i if done
  end
  i
end

File: scripts/linters/src/Linters/ContiguousUseBlock.php
<?php

declare(strict_types=1);

namespace Shirabe\Lint\Linters;

use Shirabe\Lint\Linter;
use Shirabe\Lint\Support\FileFinder;
use Shirabe\Lint\Support\Paths;

final class ContiguousUseBlock implements Linter
{
    private const USE_START_RE = '/\A(?:pub(?:\([^)]*\))?\s+)?use\b/';

    public function name(): string
    {
        return 'contiguous_use_block';
    }

    public function failureIntro(): string
    {
        return "Found blank lines splitting the leading `use` block into sections.\n"
            . 'All `use` statements at the top of the file must be contiguous (no blank lines between them):';
    }

    public function check(string $rootDir, array $excludes): array
    {
        $errors = [];

        foreach (FileFinder::rustFiles($rootDir) as $path) {
            $relative = Paths::relativeTo($rootDir, $path);
            if (in_array($relative, $excludes, true)) {
                continue;
            }

            array_push($errors, ...$this->findSplitUseBlock($path, $relative));
        }

        return $errors;
    }

    /** @return list<string> */
    private function findSplitUseBlock(string $path, string $relative): array
    {
        $lines = file($path);
        $errors = [];
        $count = count($lines);

        $i = $this->skipPreamble($lines);
        if ($i === null) {
            return [];
        }

        while (true) {
            $i = $this->consumeUseStatement($lines, $i);
            if ($i >= $count) {
                break;
            }

            $blanks = [];
            $j = $i;
            while ($j < $count) {
                $stripped = trim($lines[$j]);
                if ($stripped === '') {
                    $blanks[] = $j;
                    $j++;
                } elseif (str_starts_with($stripped, '//') || str_starts_with($stripped, '#[')) {
                    $j++;
                } else {
                    break;
                }
            }

            if ($j < $count && preg_match(self::USE_START_RE, trim($lines[$j]))) {
                foreach ($blanks as $bi) {
                    $errors[] = "{$relative}:" . ($bi + 1) . ': blank line splits the leading `use` block';
                }
                $i = $j;
            } else {
                break;
            }
        }

        return $errors;
    }

    /** @param list<string> $lines */
    private function skipPreamble(array $lines): ?int
    {
        foreach ($lines as $idx => $raw) {
            $stripped = trim($raw);
            if (preg_match(self::USE_START_RE, $stripped)) {
                return $idx;
            }
            if ($stripped === '' || str_starts_with($stripped, '//') || str_starts_with($stripped, '#![') || str_starts_with($stripped, '#[')) {
                continue;
            }

            return null;
        }

        return null;
    }

    /** @param list<string> $lines */
    private function consumeUseStatement(array $lines, int $startIdx): int
    {
        $braceDepth = 0;
        $i = $startIdx;
        $count = count($lines);

        while ($i < $count) {
            $line = $lines[$i];
            $braceDepth += substr_count($line, '{') - substr_count($line, '}');
            $done = $braceDepth <= 0 && str_ends_with(rtrim($line), ';');
            $i++;
            if ($done) {
                return $i;
            }
        }

        return $i;
    }
}
